src/main/java/com/algorand/algosdk/account/Account.java

```diff
@@ -514,22 +555,48 @@ public LogicsigSignature appendToLogicsig(LogicsigSignature lsig) throws Illegal
 
     }
 
+    public static SignedTransaction signLogicTransactionWithAddress(LogicsigSignature lsig, Address lsigAddr, Transaction tx)
+            throws IllegalArgumentException, IOException {
+        try {
+            if (!lsig.verify(lsigAddr))
+                throw new IllegalArgumentException("verification failed on logic sig");
+        } catch (NoSuchAlgorithmException ex) {
+            throw new IllegalArgumentException("verification failed on logic sig", ex);
+        }
+
+        try {
+            SignedTransaction stx = new SignedTransaction(tx, lsig, tx.txID());
+            if (!stx.tx.sender.equals(lsigAddr))
+                stx.authAddr = lsigAddr;
+            return stx;
+        } catch (Exception ex) {
+            throw new IOException("could not encode transactions", ex);
+        }
+    }
+
     /**
      * Creates SignedTransaction from LogicsigSignature and Transaction.
      * LogicsigSignature must be valid and verifiable against transaction sender field.
      * @param lsig LogicsigSignature
      * @param tx Transaction
      * @return SignedTransaction
      */
-    public static SignedTransaction signLogicsigTransaction(LogicsigSignature lsig, Transaction tx) throws IllegalArgumentException, IOException {
-        if (!lsig.verify(tx.sender)) {
-            throw new IllegalArgumentException("verification failed");
-        }
-
+    public static SignedTransaction signLogicsigTransaction(LogicsigSignature lsig, Transaction tx)
+            throws IllegalArgumentException, IOException {
+        boolean hasSig = lsig.sig != null;
+        boolean hasMsig = lsig.msig != null;
+        Address lsigAddr;
         try {
-            return new SignedTransaction(tx, lsig, tx.txID());
-        } catch (Exception ex) {
-            throw new IOException("could not encode transactions", ex);
+            if (hasSig) {
+                lsigAddr = tx.sender;
+            } else if (hasMsig) {
+                lsigAddr = lsig.msig.convertToMultisigAddress().toAddress();
+            } else {
+                lsigAddr = lsig.toAddress();
+            }
+            return Account.signLogicTransactionWithAddress(lsig, lsigAddr, tx);
+        } catch (NoSuchAlgorithmException ex) {
+            throw new IllegalArgumentException("could not sign transaction", ex);
```

Comment:
nit: can you change this to an `IOException` like before?

Reply:
Yea, need to check this